trie/branches.py

```diff
@@ -0,0 +1,167 @@
+    if node_hash == BLANK_HASH:
+        return []
+    node = db[node_hash]
+    nodetype, left_child, right_child = parse_node(node)
+    if nodetype == LEAF_TYPE:
+        if not keypath:
+            return [node]
+        else:
+            raise InvalidKeyError("Key too long")
+    elif nodetype == KV_TYPE:
+        if not keypath:
+            raise InvalidKeyError("Key too short")
+        if keypath[:len(left_child)] == left_child:
+            return [node] + _get_branch(
+                db,
+                right_child,
+                keypath[len(left_child):]
+            )
+        else:
+            raise InvalidKeyError("Key does not exist")
+    elif nodetype == BRANCH_TYPE:
+        if not keypath:
+            raise InvalidKeyError("Key too short")
+        if keypath[:1] == BYTE_0:
+            return [node] + _get_branch(db, left_child, keypath[1:])
+        else:
+            return [node] + _get_branch(db, right_child, keypath[1:])
+
+
+def verify_branch(branch, root_hash, key, value):
+    validate_is_bytes(key)
+
+    db = {keccak(node): node for node in branch}
+    assert BinaryTrie(db=db, root_hash=root_hash).get(key) == value
+    return True
+
+
+def get_trie_nodes(db, node_hash):
+    """
+    Get full trie of a given root node
+    """
+    if node_hash in db:
+        node = db[node_hash]
+    else:
+        return []
+    nodetype, left_child, right_child = parse_node(node)
+    if nodetype == KV_TYPE:
+        return [node] + get_trie_nodes(db, right_child)
+    elif nodetype == BRANCH_TYPE:
+        return [node] + get_trie_nodes(db, left_child) + get_trie_nodes(db, right_child)
+    elif nodetype == LEAF_TYPE:
+        return [node]
+
+
+def get_witness(db, node_hash, key):
+    """
+    Get all witness given a keypath prefix.
+    Include
+
+    1. witness along the keypath and
+    2. witness in the subtrie of the last node in keypath
+    """
+    validate_is_bytes(key)
+
+    return _get_witness(db, node_hash, encode_to_bin(key))
+
+
+def _get_witness(db, node_hash, keypath):
+    if not keypath:
+        return get_trie_nodes(db, node_hash)
+    if node_hash in db:
+        node = db[node_hash]
+    else:
+        return []
+    nodetype, left_child, right_child = parse_node(node)
+    if nodetype == LEAF_TYPE:
+        if keypath:
+            raise InvalidKeyError("Key too long")
+    elif nodetype == KV_TYPE:
+        if len(keypath) < len(left_child) and left_child[:len(keypath)] == keypath:
+            return [node] + get_trie_nodes(db, right_child)
+        if keypath[:len(left_child)] == left_child:
+            return [node] + _get_witness(db, right_child, keypath[len(left_child):])
+        else:
+            raise InvalidKeyError("Key does not exist")
+    elif nodetype == BRANCH_TYPE:
+        if keypath[:1] == BYTE_0:
+            return [node] + _get_witness(db, left_child, keypath[1:])
+        else:
+            return [node] + _get_witness(db, right_child, keypath[1:])
```

Comment:
Can you add a final `else` with a `raise Exception("Invariant: unreachable code path")` just for good measure.